propagate.go

```diff
@@ -0,0 +1,118 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel"
+
+	"github.com/cyverse-de/go-mod/restutils"
+	"github.com/cyverse-de/group-propagator/client/datainfo"
+	"github.com/cyverse-de/group-propagator/client/groups"
+)
+
+// To propagate a group:
+// * Fetch group details and members via iplant-groups
+//   -> get a model.GrouperGroup and model.GrouperGroupMembers, probably
+// * Determine iRODS group name (@grouper-<GrouperGroup.ID>)
+// * Create or update group with proper membership list via data-info, potentially validating users/etc.
+
+type Propagator struct {
+	groupsClient *groups.GroupsClient
+	groupPrefix  string
+
+	dataInfoClient *datainfo.DataInfoClient
+}
+
+func NewPropagator(groupsClient *groups.GroupsClient, groupPrefix string, dataInfoClient *datainfo.DataInfoClient) *Propagator {
+	if groupPrefix == "" {
+		groupPrefix = "@grouper-"
+	}
+
+	return &Propagator{
+		groupsClient:   groupsClient,
+		groupPrefix:    groupPrefix,
+		dataInfoClient: dataInfoClient,
+	}
+}
+
+func (p *Propagator) getGroupMembers(ctx context.Context, groupName string) ([]string, error) {
+	ctx, span := otel.Tracer(otelName).Start(ctx, "getGroupMembers")
+	defer span.End()
+
+	var m []string
+
+	members, err := p.groupsClient.GetGroupMembers(ctx, groupName)
+	if err != nil {
+		return m, errors.Wrapf(err, "Failed fetching Grouper group members for %s", groupName)
+	}
+
+	for _, member := range members.Members {
+		if member.SourceID == "ldap" {
+			m = append(m, member.ID)
+		} else if member.SourceID == "g:gsa" {
+			// this is a group that is a member of a group
+			submem, err := p.getGroupMembers(ctx, member.Name)
+			if err != nil {
+				return m, errors.Wrapf(err, "Failed recursing to fetch members of %s", member.Name)
+			}
+			m = append(m, submem...)
+		} else {
+			log.Errorf("Could not add group member %+v", member)
+		}
+	}
+
+	return m, nil
+}
+
+func (p *Propagator) PropagateGroupById(ctx context.Context, groupID string) error {
+	ctx, span := otel.Tracer(otelName).Start(ctx, "PropagateGroupByID")
+	defer span.End()
+
+	irodsName := fmt.Sprintf("%s%s", p.groupPrefix, groupID)
+
+	g, err := p.groupsClient.GetGroupByID(ctx, groupID)
+	if restutils.GetStatusCode(err) == 404 {
+		err = p.dataInfoClient.DeleteGroup(ctx, irodsName)
+		if err != nil {
+			err = errors.Wrap(err, "Error deleting group")
+		}
+		return err
+	} else if err != nil {
+		return errors.Wrap(err, "Failed fetching Grouper group by ID")
+	}
+
+	irodsName = fmt.Sprintf("%s%s", p.groupPrefix, g.ID)
```

Comment:
I wasn't expecting this. Is it possible for the retrieved group to have a different ID than the one we sent to the groups service?

Reply:
I don't _think_ so -- the one thing I wouldn't be sure about is what happens around renames and moves and such, but I don't think that'd mean the IDs don't line up. I think that my thinking while writing this was basically, I have to assemble this IRODS group name at some point, so I may as well use the one returned from the service. But then I added in support for the case where there isn't a group in Grouper & therefore deleting the provided ID in irods, so I had to construct the `irodsName` sooner and just changed this to reassign.

I think that I'll change this to just confirm that the returned group ID matches the one from the message and throw an error if it doesn't, and we can keep some sort of an eye on it.